crates/handlers/src/upstream_oauth2/logout.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright 2025 New Vector Ltd.
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+// Please see LICENSE in the repository root for full details.
+
+use mas_axum_utils::cookies::CookieJar;
+use mas_router::UrlBuilder;
+use mas_storage::{RepositoryAccess, upstream_oauth2::UpstreamOAuthProviderRepository};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tracing::{error, warn};
+use url::Url;
+
+use super::UpstreamSessionsCookie;
+use crate::impl_from_error_for_route;
+
+#[derive(Serialize, Deserialize)]
+struct LogoutToken {
+    logout_token: String,
+}
+
+/// Structure to collect upstream RP-initiated logout endpoints for a user
+#[derive(Debug, Default)]
+pub struct UpstreamLogoutInfo {
+    /// Collection of logout endpoints that the user needs to be redirected to
+    pub logout_endpoints: String,
+    /// Optional post-logout redirect URI to come back to our app
+    pub post_logout_redirect_uri: Option<String>,
+}
+
+#[derive(Debug, Error)]
+pub enum RouteError {
+    #[error(transparent)]
+    Internal(Box<dyn std::error::Error + Send + Sync + 'static>),
+
+    #[error("provider was not found")]
+    ProviderNotFound,
+
+    #[error("session was not found")]
+    SessionNotFound,
+}
+
+impl_from_error_for_route!(mas_storage::RepositoryError);
+
+impl From<reqwest::Error> for RouteError {
+    fn from(err: reqwest::Error) -> Self {
+        Self::Internal(Box::new(err))
+    }
+}
+
+/// Get RP-initiated logout URLs for a user's upstream providers
+///
+/// This retrieves logout endpoints from all connected upstream providers that
+/// support RP-initiated logout.
+///
+/// # Parameters
+///
+/// * `repo`: The repository to use
+/// * `url_builder`: URL builder for constructing redirect URIs
+/// * `cookie_jar`: Cookie from user's browser session
+///
+/// # Returns
+///
+/// Information about upstream logout endpoints the user should be redirected to
+///
+/// # Errors
+///
+/// Returns a `RouteError` if there's an issue accessing the repository
+pub async fn get_rp_initiated_logout_endpoints<E>(
+    url_builder: &UrlBuilder,
+    repo: &mut impl RepositoryAccess<Error = E>,
+    cookie_jar: &CookieJar,
+) -> Result<UpstreamLogoutInfo, RouteError>
+where
+    RouteError: std::convert::From<E>,
+{
+    let mut result: UpstreamLogoutInfo = UpstreamLogoutInfo::default();
+    // Set the post-logout redirect URI to our app's logout completion page
+    let post_logout_redirect_uri = url_builder
+        .absolute_url_for(&mas_router::Login::default())
+        .to_string();
+    result.post_logout_redirect_uri = Some(post_logout_redirect_uri.clone());
+
+    let sessions_cookie = UpstreamSessionsCookie::load(cookie_jar);
```

Comment:
Loading from the cookie is unreliable. You should instead:

 - lookup the authentications for the current `user_session` in the `user_session_authentications`
 - authentications that were done through an upstream will have a reference to a `upstream_oauth_authorization_session_id`
 - there you can grab the `id_token` received during the auth and extract the subject
 - you should also pass the id_token to the RP-initiated logout endpoint through the `id_token_hint`

Reply:
Thank you for your review!
I have amend the code but I was not sure if I need to extract the subject to perform the logout